Reload Swift Package when new file creation is indicated by `DidChangeWatchedFileNotification`

ahoppen opened the pull request:
Implement rudementary support for `DidChangeWatchedFileNotification` for SwiftPM projects: When a file is added, reload the Swift package to compute build settings for the new file.

This enables proper semantic functionality added to the project after the LSP server was started.

Resolves rdar://87149128 [SR-15633](https://bugs.swift.org/browse/SR-15633).

ahoppen commented:
@swift-ci Please test

benlangmuir commented on Sources/SKSwiftPMWorkspace/SwiftPMWorkspace.swift:
```diff
@@ -293,6 +293,13 @@ extension SwiftPMWorkspace: SKCore.BuildSystem {
+  public func filesDidChange(_ events: [FileEvent]) {
+    if events.contains(where: { $0.type == .created }) {
+      // TODO: It should not be necessary to reload the entire package just to get build settings for one file.
+      try? self.reloadPackage()
```
Instead of `try?` you could wrap in `orLog` so that the failure is logged.

DavidGoldman replied:
Will things be in an invalid state if we set self.packageGraph but https://github.com/apple/sourcekit-lsp/blob/8503a3fc957118d6bf4b7f8ca687408e05048ddf/Sources/SKSwiftPMWorkspace/SwiftPMWorkspace.swift#L170  fails? Might want to set the packageGraph after that, Ben might know more.

ahoppen replied:
> Instead of `try?` you could wrap in `orLog` so that the failure is logged.

Oh, nice, I didn’t know about `orLog`. Thanks.

> Will things be in an invalid state if we set self.packageGraph but
> 
> https://github.com/apple/sourcekit-lsp/blob/8503a3fc957118d6bf4b7f8ca687408e05048ddf/Sources/SKSwiftPMWorkspace/SwiftPMWorkspace.swift#L170
> 
> fails? Might want to set the packageGraph after that, Ben might know more.

Good point. I think we should be safe if we create `BuildPlan` before setting `packageGraph`. I’ve flipped the two statement. What do you think?

DavidGoldman replied:
LGTM

benlangmuir commented on Sources/SourceKitLSP/SourceKitServer.swift:
```diff
@@ -582,6 +583,14 @@ extension SourceKitServer {
+    /// Request the client to send us DidChangeWatchedFileNotifications for all new files.
+    /// Currently, these notifications are only handled by SwiftPMWorkspace, which will reload the package when a new file is added.
+    let watchers = [
+      FileSystemWatcher(globPattern: "**", kind: [.create])
```
How about
* Also handle `.delete` on all files
* Add a pattern for `Package.swift` (or maybe `**/Package.swift`?) with `.change`

DavidGoldman replied:
Should this be in SourceKitServer though or should we let the BuildSystem tell us what we should watch?

benlangmuir replied:
I'm fine with doing it either way in this PR.  You're right that in principle there should be interaction with the build system for this.  I'm not sure what that API should actually look like though, because in principle the build system also wants to watch paths that might be outside the workspace -- e.g. the build directory may not be in the workspace -- and couldn't necessarily rely on the LSP notifications for those specific paths. So I would say do the simple thing and make it a fixed list either here or as a property on the build system :shrug:

ahoppen replied:
> * Also handle `.delete` on all files

I thought about the same but don’t think it’s necessary. There are no new build settings to complete and the old ones, which reference the deleted file, still seem to work for sourcekitd.

> * Add a pattern for `Package.swift` (or maybe `**/Package.swift`?) with `.change`

I’d like to check how many notifications VSCode sends us if listen to change events as well. If we receive a change notification for every keystroke and all keystrokes trigger a complete package reload, this might be too performance critical. Let’s do this in a follow-up PR.

> Should this be in SourceKitServer though or should we let the BuildSystem tell us what we should watch?

Since we are not really narrowing down on any specific pattern here, I don’t think there’s a point in letting the `BuildSystem` tell us what to watch for, it could just ignore the events it doesn’t care for. I agree that it should be a `BuildSystem` property if we only listened for changes to `.swift` files, but let’s leave this as an optimization for follow-up PRs, I’d say.

benlangmuir replied:
> I’d like to check how many notifications VSCode sends us if listen to change events as well. If we receive a change notification for every keystroke and all keystrokes trigger a complete package reload, this might be too performance critical. Let’s do this in a follow-up PR.

It would not be correct to send changes per keystroke, these APIs are only for changes to the file, not to the in-memory text buffer.  We can do it in a follow-up PR if you like, but I think this is important before we close the SR issue.

benlangmuir replied:
Or for the SR, you could un-dupe https://bugs.swift.org/browse/SR-15679. That would be fine too.

benlangmuir replied:
RE: delete

It's great that sourcekit is being lenient about missing files, but I think we should still strive to use accurate settings.
* clangd: We use the index to choose a main file for a header, and if that main file is removed, we need to update our settings to use a different main file if possible.  This should already work in the build system manager except that it needs to get the `.removedOrUnavailable` update from the underlying build system.
* swiftpm: We're implicitly relying on swiftpm to not make any *other* changes to settings based on the removal of a file except removing it from the module sources. That may be true today, but it could change.
* other uses of settings: We shouldn't assume we won't use these settings for other things like invoking the compiler in the future, and it will fail if we're missing an input.

ahoppen replied:
Sounds good. I’ve added `.delete`.

benlangmuir commented on Sources/SKCore/BuildSystemManager.swift:
```diff
@@ -125,6 +125,11 @@ public final class BuildSystemManager {
   }
+
+  public func filesDidChange(_ events: [FileEvent]) {
+    buildSystem?.filesDidChange(events)
```
This should probably be wrapped in `queue.async`. Otherwise we are doing settings queries on the build system manager queue, but triggering file changes on the main server queue, which is racy.  Alternatively we could make the SwiftPM build system itself thread-safe, but that seems like more boilerplate.  Hopefully we can switch to swift concurrency and it won't be so annoying.

ahoppen replied:
Also running the `BuildSystem.filesDidChange` code async is beneficial I think so we don’t block the queue that receives the LSP messages.

benlangmuir reviewed (commented):
`reloadPackages` needs to be taught to notify the build system delegate of the settings changes using `delegate.fileBuildSettingsChanged` for all the registered URLs (and we'll need to start keeping a list of the registered URLs in the SwiftPM build system).  That will fix the settings in files that were already open before the change.

Excited to see this happening!

benlangmuir commented:
Also, don't forget to remove this [from the README](https://github.com/apple/sourcekit-lsp#caveats):
> SwiftPM build settings are not updated automatically after files are added/removed.
> * Workaround: close and reopen the project after adding/removing files

😸

DavidGoldman commented on Sources/SourceKitLSP/CapabilityRegistry.swift:
```diff
@@ -82,6 +85,28 @@ public final class CapabilityRegistry {
+    watchers: [FileSystemWatcher],
+    registerOnClient: ClientRegistrationHandler
+  ) {
+    guard clientHasDynamicCompletionRegistration else { return }
```
This should be for didchangedwatchedfiles?

DavidGoldman commented on Sources/SKSwiftPMWorkspace/SwiftPMWorkspace.swift:
```diff
@@ -293,6 +293,13 @@ extension SwiftPMWorkspace: SKCore.BuildSystem {
+
+  public func filesDidChange(_ events: [FileEvent]) {
+    if events.contains(where: { $0.type == .created }) {
+      // TODO: It should not be necessary to reload the entire package just to get build settings for one file.
```
Is this true? I'm not sure exactly how the SPM integration works, is there some incremental reload that we could do?

ahoppen replied:
I’m not sure if it’s currently possible with the API SwiftPM provides, but in principle it shouldn’t be necessary to reload the entire package just to get the build settings for one new file.

DavidGoldman commented on Tests/SourceKitLSPTests/SwiftPMIntegration.swift:
```diff
@@ -56,4 +56,67 @@ final class SwiftPMIntegrationTests: XCTestCase {
+
+    let newFile = ws.testLoc("newFile:call")
+
+    // Check that we don't get cross-file code completion befor we sent a `DidChangeWatchedFilesNotification` to make sure we didn't include the file in the initial retrieval of build settings.
```
before we send

DavidGoldman commented:
> `reloadPackages` needs to be taught to notify the build system delegate of the settings changes using `delegate.fileBuildSettingsChanged` for all the registered URLs (and we'll need to start keeping a list of the registered URLs in the SwiftPM build system). That will fix the settings in files that were already open before the change.
> 
> Excited to see this happening!

Yep, this is great! :D Thanks for working on it

ahoppen commented:
Thanks for the feedback @benlangmuir and @DavidGoldman. I addressed the inline comments by squashing the first commit and introduced notification of the build system delegate on package reloading in a second commit.

The points still open are:
- Should we also listen for deleted files? (https://github.com/apple/sourcekit-lsp/pull/443#discussion_r781367077)
- Watch for changes to `Package.swift`, which I’ll add in a follow-up PR (https://github.com/apple/sourcekit-lsp/pull/443#discussion_r780438117)

ahoppen commented:
@swift-ci Please test

ahoppen commented on Sources/SKSwiftPMWorkspace/SwiftPMWorkspace.swift:
```diff
@@ -293,6 +323,35 @@ extension SwiftPMWorkspace: SKCore.BuildSystem {
   }
+
+  /// An event is relevant if it modifies a file that matches one of the file rules used by the SwiftPM workspace.
+  private func isFileEventRelevant(event: FileEvent) -> Bool {
+    // FIXME: fileRules should also include self.workspace.additionalFileRules and toolsVersion should be self.workspace.currentToolsVersion
+    let fileRules = FileRuleDescription.builtinRules
+    let toolsVersion = ToolsVersion.currentToolsVersion
+
+    guard let fileURL = event.uri.fileURL else {
+      return false
+    }
+    let filePath = AbsolutePath(fileURL.path)
+    switch event.type {
+    case .created, .deleted:
+      return fileRules.contains(where: { $0.match(path: filePath, toolsVersion: toolsVersion) })
+    case .changed:
+      return false
+    default: // Unknown file change type
+      return false
+    }
+  }
```
@neonichu Could you take a look whether this code makes sense? The intention is to match only files which might impact the build settings computed by SwiftPM. If you think this is correct, I’ll open a PR on SwiftPM to make `Workspace.additionalFileRules` and `Workspace.currentToolsVersion` public.

neonichu replied:
I think it might be better to find a way to refactor `TargetSourcesBuilder` in SwiftPM to provide this kind of information instead of interacting with file rules directly.

ahoppen replied:
Do you have any particular path in mind how `Workspace` should expose the `TargetSourcesBuilder`? AFAICT it only creates an ad-hoc `PackageBuilder`, which in turn creates an ad-hoc `TargetSourcesBuilder`.

neonichu replied:
That's a good question, do you have an existing package graph at this point from which you could get a `TargetDescription` to pass down? If yes, maybe just a new function on `Workspace` that gets package identity etc and the target description and returns a configured `TargetSourcesBuilder`? Additionally, I think target sources builder would need a function that would decide whether a singular file path is part of the given package or not.

neonichu replied:
Ah you're right, I wonder whether we could change `TargetSourcesBuilder` to take a `ResolvedTarget` instead. This is turning out to be a bit of a bigger change than I originally thought.

I guess another option could be that we add a new function to `Workspace` to which you pass all the info you have and for now it would just do a simple check using file rules, but we could later refine it to be more accurate. That way we can decouple your PR from making bigger changes to SwiftPM.

BTW, `ManifestJSONParser.parseTarget` is the general purpose target creation, because it is just done as part of manifest loading. We run the compiled manifest, it produces JSON and then we load that JSON.

ahoppen replied:
After some offline discussion with @neonichu, we have decided to only watch files matching extensions specified in `FileRuleDescription.builtinRules` in the client. Additionally, I’m adding a function to SwiftPM that will allow more granular filtering based on e.g. the directory structure of a package in the future: https://github.com/apple/swift-package-manager/pull/4016

DavidGoldman commented on Sources/SKSwiftPMWorkspace/SwiftPMWorkspace.swift:
```diff
@@ -293,6 +323,35 @@ extension SwiftPMWorkspace: SKCore.BuildSystem {
+    if events.contains(where: { isFileEventRelevant(event: $0) }) {
+      // TODO: It should not be necessary to reload the entire package just to get build settings for one file.
+      orLog {
+        try self.reloadPackage()
```
Do we need to do this on self.queue as well?

ahoppen replied:
It doesn’t need to be, because `BuildSystemManager` is already guarding against race conditions in `BuildSystemManager.filesDidChange` by informing the build systems about file changes serially on its own queue.

DavidGoldman replied:
I think it does since `registerForChangeNotifications` (on queue) could be called right before `filesDidChange`, and then `reloadPackage` (off queue) could mutate `packageGraph`/`fileToTarget`/`sourceDirToTarget ` while `registerForChangeNotifications` is reading it on the queue, right? So we need some way to synchronize that.

ahoppen replied:
Oh, yes, we also need to guard `packageGraph`, `fileToTarget` and `sourceDirToTarget` against race conditions. I only thought of `watchedFiles` and `delegate` before. Added the guards in https://github.com/apple/sourcekit-lsp/pull/443/commits/2e3b78312f40b89ba1c15f39a062c1881ad89e24.

ahoppen commented:
https://github.com/apple/swift-package-manager/pull/4016

@swift-ci Please test

ahoppen commented:
apple/swift-package-manager#4016

@swift-ci Please test

benlangmuir commented on Tests/SourceKitLSPTests/SwiftPMIntegration.swift:
```diff
@@ -56,4 +56,88 @@ final class SwiftPMIntegrationTests: XCTestCase {
     ])
   }
+
+  func testAddFile() throws {
```
Need to update Linux tests.  Incidentally, I wonder if we could switch to automatic test discovery on Linux at this point - it's been around for a while now.

ahoppen replied:
Thanks for catching this. I’ll try to switch to automatic test discovery in a follow-up PR.

ahoppen commented on Sources/SKSwiftPMWorkspace/SwiftPMWorkspace.swift:
```diff
@@ -58,6 +58,18 @@ public final class SwiftPMWorkspace {
+  /// - `packageGraph`
+  /// - `fileToTarget`
+  /// - `sourceDirToTarget`
+  let queue: DispatchQueue = .init(label: "SwiftPMWorkspace.queue", qos: .background)
```
I wonder how much of a difference it makes in practice. But yes, the description of `.utility` does seem to match more closely to what we are doing.

ahoppen commented:
apple/swift-package-manager#4016

@swift-ci Please test

benlangmuir commented on Sources/SourceKitLSP/SourceKitServer.swift:
```diff
@@ -21,6 +21,7 @@ import SKSupport
 import TSCBasic
 import TSCLibc
 import TSCUtility
+import PackageLoading
```
It would be nice if we could contain all imports of SwiftPM modules to the SwiftPM build system file to keep things contained.  No need to change it in this PR though.

ahoppen commented:
apple/swift-package-manager#4016

@swift-ci Please test
